src/persistent_backend_manager.h

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2023, NVIDIA CORPORATION. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <list>
+#include "backend_manager.h"
+#include "constants.h"
+#include "triton/common/model_config.h"
+#include "status.h"
+
+namespace triton { namespace core {
+
+//
+// A single manager across all InferenceServer objects that keeps the
+// persistent backends loaded as long as there is at least one
+// InferenceServer object.
+//
+class PersistentBackendManager {
+ public:
+  static Status Create(
+      const triton::common::BackendCmdlineConfigMap& config_map,
+      std::shared_ptr<PersistentBackendManager>* manager, std::shared_ptr<TritonBackendManager> backend_manager);
+
+ private:
+  DISALLOW_COPY_AND_ASSIGN(PersistentBackendManager);
+  PersistentBackendManager(std::shared_ptr<TritonBackendManager> backend_manager) : backend_manager_(backend_manager) {
+  }
+  Status InitPersistentBackends(const triton::common::BackendCmdlineConfigMap& config_map);
+  Status InitPersistentBackend(
+      const std::string& backend_name,
+      const triton::common::BackendCmdlineConfigMap& config_map);
+
+  std::list<std::shared_ptr<TritonBackend>> persist_backends_;
+  std::shared_ptr<TritonBackendManager> backend_manager_;
```

Comment:
Is there a reason we need to keep a point to the backend manager? We should avoid these doubly links when possible.

Reply:
Wait, is this the whole reason for this class, to keep and hold onto the pointer? That is a bit of a hack :-/

Reply:
This code is mostly ported from an older Triton release: https://github.com/triton-inference-server/server/blob/r22.02/src/core/persistent_backend_manager.cc

I have updated the code to not store the `backend_manager_` as an attribute.

> Wait, is this the whole reason for this class, to keep and hold onto the pointer? That is a bit of a hack :-/

This class is mostly a singleton designed to load a set of backends before any backend/model is loaded.